File: src/org/cappuccino_project/ide/intellij/plugin/stubs/types/ObjJPreprocessorDefineFunctionStubType.java
package org.cappuccino_project.ide.intellij.plugin.stubs.types;

import com.intellij.psi.stubs.StubElement;
import org.cappuccino_project.ide.intellij.plugin.psi.impl.ObjJPreprocessorDefineFunctionImpl;
import org.cappuccino_project.ide.intellij.plugin.stubs.impl.ObjJPreprocessorFunctionDeclarationStubImpl;
import org.cappuccino_project.ide.intellij.plugin.stubs.interfaces.ObjJFunctionDeclarationElementStub;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public class ObjJPreprocessorDefineFunctionStubType extends ObjJAbstractFunctionDeclarationStubType<ObjJPreprocessorDefineFunctionImpl, ObjJPreprocessorFunctionDeclarationStubImpl> {

    ObjJPreprocessorDefineFunctionStubType(
            @NotNull
                    String debugName) {
        super(debugName, ObjJPreprocessorDefineFunctionImpl.class, ObjJPreprocessorFunctionDeclarationStubImpl.class);
    }

    @Override
    public ObjJPreprocessorDefineFunctionImpl createPsi(
            @NotNull
                    ObjJFunctionDeclarationElementStub<ObjJPreprocessorDefineFunctionImpl> stub) {
        return new ObjJPreprocessorDefineFunctionImpl(stub, ObjJStubTypes.PREPROCESSOR_FUNCTION);
    }

    @NotNull
    @Override
    ObjJFunctionDeclarationElementStub<ObjJPreprocessorDefineFunctionImpl> createStub(StubElement parent,
                                                                                 @NotNull
                                                                                         String fileName,
                                                                                 @NotNull
                                                                                         String fqName,
                                                                                 @NotNull
                                                                                         List<String> paramNames,
                                                                                 @Nullable
                                                                                         String returnType) {
        return new ObjJPreprocessorFunctionDeclarationStubImpl(parent, fileName, fqName, paramNames, returnType);
    }
}

File: src/org/cappuccino_project/ide/intellij/plugin/stubs/types/ObjJPreprocessorDefineFunctionStubType.kt
package org.cappuccino_project.ide.intellij.plugin.stubs.types

import com.intellij.psi.stubs.StubElement
import org.cappuccino_project.ide.intellij.plugin.psi.impl.ObjJPreprocessorDefineFunctionImpl
import org.cappuccino_project.ide.intellij.plugin.stubs.impl.ObjJPreprocessorFunctionDeclarationStubImpl
import org.cappuccino_project.ide.intellij.plugin.stubs.interfaces.ObjJFunctionDeclarationElementStub

class ObjJPreprocessorDefineFunctionStubType internal constructor(
        debugName: String) : ObjJAbstractFunctionDeclarationStubType<ObjJPreprocessorDefineFunctionImpl, ObjJPreprocessorFunctionDeclarationStubImpl>(debugName, ObjJPreprocessorDefineFunctionImpl::class.java, ObjJPreprocessorFunctionDeclarationStubImpl::class.java) {

    override fun createPsi(
            stub: ObjJFunctionDeclarationElementStub<ObjJPreprocessorDefineFunctionImpl>): ObjJPreprocessorDefineFunctionImpl {
        return ObjJPreprocessorDefineFunctionImpl(stub, ObjJStubTypes.PREPROCESSOR_FUNCTION)
    }

    internal override fun createStub(parent: StubElement<*>,
                                     fileName: String,
                                     fqName: String,
                                     paramNames: List<String>,
                                     returnType: String?,
                                     shouldResolve: Boolean): ObjJFunctionDeclarationElementStub<ObjJPreprocessorDefineFunctionImpl> {
        return ObjJPreprocessorFunctionDeclarationStubImpl(parent, fileName, fqName, paramNames, returnType, shouldResolve)
    }
}
